solidity/contracts/bridge/Wallets.sol

```diff
@@ -0,0 +1,287 @@
+    }
+
+    /// @notice Requests creation of a new wallet. This function just
+    ///         forms a request and the creation process is performed
+    ///         asynchronously. Outcome of that process should be delivered
+    ///         using `registerNewWallet` function.
+    /// @param activeWalletMainUtxo Data of the active wallet's main UTXO, as
+    ///        currently known on the Ethereum chain.
+    /// @dev Requirements:
+    ///      - `activeWalletMainUtxo` components must point to the recent main
+    ///        UTXO of the given active wallet, as currently known on the
+    ///        Ethereum chain. If there is no active wallet at the moment, or
+    ///        the active wallet has no main UTXO, this parameter can be
+    ///        empty as it is ignored.
+    ///      - Wallet creation must not be in progress
+    ///      - If the active wallet is set, one of the following
+    ///        conditions must be true:
+    ///        - The active wallet BTC balance is above the minimum threshold
+    ///          and the active wallet is old enough, i.e. the creation period
+    ///           was elapsed since its creation time
+    ///        - The active wallet BTC balance is above the maximum threshold
+    ///        If the active wallet is not set at the moment, there is no
+    ///        additional conditions.
+    function requestNewWallet(
+        Data storage self,
+        BitcoinTx.UTXO calldata activeWalletMainUtxo
+    ) external {
+        // TODO: Uncomment when `getWalletCreationState` will be exposed by
+        //       EcdsaWalletRegistry.
+        //
+        // require(
+        //     self.registry.getWalletCreationState() == EcdsaDkg.State.IDLE,
+        //     "Wallet creation already in progress"
+        // );
+
+        bytes20 activeWalletPubKeyHash = self.activeWalletPubKeyHash;
+
+        // If the active wallet is set, fetch this wallet's details from
+        // storage to perform conditions check. The `registerNewWallet`
+        // function guarantees an active wallet is always one of the
+        // registered ones.
+        if (activeWalletPubKeyHash != bytes20(0)) {
+            uint64 activeWalletBtcBalance = getWalletBtcBalance(
+                self,
+                activeWalletPubKeyHash,
+                activeWalletMainUtxo
+            );
+            uint32 activeWalletCreatedAt = self
+                .registeredWallets[activeWalletPubKeyHash]
+                .createdAt;
+            /* solhint-disable-next-line not-rely-on-time */
+            bool activeWalletOldEnough = block.timestamp >=
+                activeWalletCreatedAt + self.creationPeriod;
+
+            require(
+                (activeWalletOldEnough &&
+                    activeWalletBtcBalance >= self.minBtcBalance) ||
+                    activeWalletBtcBalance >= self.maxBtcBalance,
+                "Wallet creation conditions are not met"
+            );
+        }
```

Comment:
We could expose functions allowing to evaluate if DKG is possible:
- `isWalletCreationPossible`
- `getWalletBtcBalance` (already there, let's just expose from the `Bridge`)
- `getWalletAge`
- plus something that would let evaluate the active wallet address

Reply:
Hmm, I'd be careful with that giving the contract size. For now we have:
- `getRegisteredWallet` that returns the entire `Wallets.Wallet` struct. Based on that we can directly read the wallet age (`createdAt`) and the BTC balance indirectly since we have `mainUtxoHash`. To read the balance the caller needs to know the plain-text main UTXO. I think we have no easy way to get it now apart from scanning the BTC chain by taking the TX hash from appropriate `DepositsSwept` and  `RedemptionsCompleted` events. If we want to make it easier, we can think about placing output index and value of the new main UTXO in those events, at a cost of increased gas consumption.
- `getActiveWalletPubKeyHash` that returns the 20-byte PKH of the wallet currently set as active. This should be enough to get the BTC address and construct the P2SH deposit script.

Regarding `isWalletCreationPossible` we can check that by calling `IWalletRegistry.getWalletCreationState` to check whether DKG is currently in progress and then check the active wallet age and balance (using methods I mentioned above) to determine the truthfulness of wallet creation conditions.

Do you think we need to add something anyway?

Reply:
I wrote this comment when just `Wallet.sol` was implemented and now looking at `Bridge.sol` I think we have all we need. One addition, out of scope of this PR, could be a `BridgeInfo` contract with view function allowing to evaluate whether wallet creation is possible to make the client implementation simpler. But since this would be just a view, let's not bother about it now.